### DISCUSSION TITLE
What does deleting instances look like?

### DISCUSSION THREAD

**Discussion by TheColonel2688** [DevrexLabs/memstate#118] [Q&A] [answered]
I've read over the quick start guide and the main read me, but I don't see anything about deleting instances.

What does deleting instances look like? Can you provide an example?

**Comment by rofr** [comment, marked as answer]
Hi there! I'm assuming you mean deleting entities or objects from your in-memory model. So let's extend the [quick start guide](https://memstate.io/docs/core-1.0/getting-started/quick-start-guide/) with a command to remove tasks:
```csharp
[Serializable]
public class RemoveTaskCommand : Command<TaskModel>
{
  public readonly int TaskIndex {get;set;}

  public override void Execute(TaskModel model)
  {
      model.Tasks.RemoveAt(TaskIndex);
  }
}
```

Removing based on the position in the list is not great design. A sounder approach would be to assign each task a unique id and remove by id instead. But hopefully this is enough to get you going. Using the command would look something like this:

```csharp
var engine = Engine.For<TaskModel>();
await engine.Execute(new RemoveTaskCommand{TaskIndex = 2});
```

Note that the quickstart referenced above is based on origodb, the predecessor of memstate. But with some minor tweaks should work with memstate as well. Let us know if you need more guidance.

**Reply by TheColonel2688** [reply]
Yes, "entities or objects from your in-memory model"

What if there is another entity (say a user) that references that task instance? will the references be removed from that user entity as well or will I have to do a separate removal and look through all the users that might reference that task, then remove it there too?

**Reply by rofr** [reply]
Great question! The example above only removes the reference from the list of tasks. If there are any other references to the object from within the model, it will never be garbage collected.  There is no magic involved, so just use your knowledge and about how objects are allocated on the heap and are eligible for recycling when there are no longer any references.

**Reply by TheColonel2688** [reply]
Ok, thank you. I think that answers my questions for now.
